### PR TITLE
Build examples using only one CI instance

### LINKED ISSUE #24: Run examples in one instance of CI server
Right now, every example runs in a separate CI server:

- This is very slow, 60% or more of the time in each one is invested in provisioning the server
- It is tedious, every new example requires to augment the scripts.

I recommend:

- Having some test script that walks the repo. For every folder that contains a "build.bat", "build.sh" or "build.py" file, consider it a test to be run
- For every test, make sure to create a temporary cache with CONAN_USER_HOME
- Accumulate the results to show a reasonable summary at the end, only exit with error after running all of them

### PR DESCRIPTION
**Pull request by uilianries**
closes #24

### CODE REVIEW

**Comment by danimtb** [comment]
Looks like the output in appveyor is being blocke and printed all at the end

```
================================================================
* FEATURES\CMAKE\FIND_PACKAGE\EXPORTED_TARGETS_MULTICONFIG
================================================================
================================================================
* FEATURES\CMAKE\FIND_PACKAGE\FIND_CMAKE_MULTI_GENERATOR_TARGETS
================================================================
================================================================
* FEATURES\EMSCRIPTEN
================================================================
================================================================
* FEATURES\LOCKFILES\CI
================================================================
================================================================
* FEATURES\MULTI_CONFIG
================================================================
================================================================
* FEATURES\PACKAGE_DEVELOPMENT_FLOW
================================================================
================================================================
* FEATURES\PYTHON_REQUIRES\REUSE_CONANFILE
================================================================
================================================================
* FEATURES\WORKSPACE\CMAKE
================================================================
================================================================
* LIBRARIES\DEAR-IMGUI\BASIC
================================================================
================================================================
* LIBRARIES\FOLLY\BASIC
================================================================
================================================================
* LIBRARIES\POCO\MD5
================================================================
================================================================
* LIBRARIES\PROTOBUF\SERIALIZATION
================================================================
=== CONAN EXAMPLES: RESULTS ===
FEATURES\CMAKE\FIND_PACKAGE\EXPORTED_TARGETS_MULTICONFIG: OK
FEATURES\CMAKE\FIND_PACKAGE\FIND_CMAKE_MULTI_GENERATOR_TARGETS: OK
FEATURES\EMSCRIPTEN: OK
FEATURES\LOCKFILES\CI: OK
FEATURES\MULTI_CONFIG: OK
FEATURES\PACKAGE_DEVELOPMENT_FLOW: OK
FEATURES\PYTHON_REQUIRES\REUSE_CONANFILE: OK
FEATURES\WORKSPACE\CMAKE: OK
LIBRARIES\DEAR-IMGUI\BASIC: OK
LIBRARIES\FOLLY\BASIC: OK
LIBRARIES\POCO\MD5: OK
LIBRARIES\PROTOBUF\SERIALIZATION: OK
```
Do you know how to fix it? I think CPT does it right, maybe we can check there. btw, I really like how the new output looks in Travis!!

**Comment by uilianries** [comment]
CPT uses sys.stdout.write + flush. I think the problem is forcing "flush" after printing something.

I'll provide a new fix ASAP.

**Comment by uilianries** [comment]
@danimtb I think it's much better now.

File: features/emscripten/build.bat [hunk 1]
```diff
@@ -1,4 +1,6 @@
 @echo off
+set CONAN_USER_HOME=%USERPROFILE%
```
Comment (uilianries) [inline features/emscripten/build.bat]: Windows Max Path :trollface:

File: appveyor.yml [hunk 2]
```diff
@@ -1,80 +1,14 @@
...
-        - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
-          CMAKE_GENERATOR: "Visual Studio 15 2017 Win64"
-          PROJECT_DIR: features/multi_config
-        - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
-          PROJECT_DIR: features/lockfiles/ci
-        # Integrate a build-system in Conan
-        - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
-          CMAKE_GENERATOR: "Visual Studio 15 2017 Win64"
-          PROJECT_DIR: features/integrate_build_system
+    PYTHON: "C:\\Python27"
```
Comment (uilianries) [inline appveyor.yml]: For some reason, Python3 can't load Conan generator for WAF: https://ci.appveyor.com/project/ConanOrgCI/examples/builds/26208045?fullLog=true#L1910

Of course, it's related to the fact that I'm running Python -> Batch -> Python -> Conan

**Comment by uilianries** [comment]
It's working now, but I think this PR needs to be squashed before merging. There is an option squash+merge on Github.

File: appveyor.yml (hunk 2 above)
Comment (jgsogo) [reply appveyor.yml]: What about using the value of `sys.executable` instead of just `python`?

File: .ci/run.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+def get_build_list():
+    builds = []
+    script = "build.bat" if platform.system() == "Windows" else "build.sh"
+    for root, dirs, files in os.walk("."):
+        if "build.py" in [os.path.basename(file) for file in files]:
+            builds.append(os.path.join(root, "build.py"))
+            continue
+
+        for file in files:
+            if os.path.basename(file) == script:
```
Comment (jgsogo) [inline .ci/run.py]:
```suggestion
            if os.path.basename(file) in [script, 'build.py']:
                builds.append(os.path.join(root, file))
                break  # End the for loop
```

and you can remove the previous `if` block

File: .ci/run.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+def print_build(script):
+    dir_name = os.path.dirname(script)
+    dir_name = dir_name[2:] if dir_name.startswith('.') else dir_name
+    writeln_console("================================================================")
+    writeln_console("* " + colorama.Style.BRIGHT + "{}".format(dir_name.upper()))
+    writeln_console("================================================================")
+
+
+def run_scripts(scripts):
+    results = {}
```
Comment (jgsogo) [inline .ci/run.py]: We may initialize the dictionary so we get the table with all the examples at the bottom, and keep the order of the scripts in the table:

```suggestion
    results = OrderedDict.fromkeys(scripts, '')
```

File: appveyor.yml (hunk 2 above)
Comment (uilianries) [reply appveyor.yml]: sounds much better

File: .ci/run.py (hunk 3 above)
Comment (uilianries) [reply .ci/run.py]: done

File: .ci/run.py (hunk 4 above)
Comment (uilianries) [reply .ci/run.py]: done

File: .ci/run.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        yield
+    finally:
+        logging.debug("cd {}".format(current))
+        os.chdir(current)
+
+
+def writeln_console(message):
+    sys.stdout.write(message)
+    sys.stdout.write('\n')
+    sys.stderr.flush()
```
[the remark this answers is not in the document]
Comment (uilianries) [reply .ci/run.py]: done

File: appveyor.yml (hunk 2 above)
Comment (uilianries) [reply appveyor.yml]: done

File: features/emscripten/build.bat (hunk 1 above)
Comment (uilianries) [reply features/emscripten/build.bat]: Moved to `C:/projects/CONAN_HOME/<example-name>`

**Comment by uilianries** [comment]
Hi @jgsogo I think it's better now, please take a look again

File: features/lockfiles/ci/build.py [hunk 6]
```diff
@@ -58,11 +58,10 @@ def rm(path):
     _, pkg_ref = build_order[0][0]
     pkg_ref = pkg_ref.split("#", 1)[0]
     print("\n********** Rebuild affected package: %s ***********" % pkg_ref)
-    run("conan graph clean-modified release/")
```
Comment (jgsogo) [inline features/lockfiles/ci/build.py]: This change is not for this PR.

File: libraries/protobuf/serialization/main.py [hunk 7]
```diff
@@ -7,7 +7,7 @@
         print("Retrieve Sensor object from sensor.data")
         sensor = Sensor()
         sensor.ParseFromString(content)
-        print(f"Sensor name: {sensor.name}")
-        print(f"Sensor temperature: {sensor.temperature}")
-        print(f"Sensor humidity: {sensor.humidity}")
+        print("Sensor name: {}".format(sensor.name))
```
Comment (jgsogo) [inline libraries/protobuf/serialization/main.py]: Now we are using py3, right? I'm not asking you to change this back to f-string unless you want to... this was your example ;D

File: features/emscripten/build.bat (hunk 1 above)
Comment (jgsogo) [reply features/emscripten/build.bat]: Are you keeping this? Is it not possible to use the `C:/projects/CONAN_HOME/<example-name>` path?

File: features/lockfiles/ci/build.py (hunk 6 above)
Comment (uilianries) [reply features/lockfiles/ci/build.py]: I know, but Conan 1.18.0 is ready and without this change will break this PR. OR, we could create a new PR only with this change.

File: libraries/protobuf/serialization/main.py (hunk 7 above)
Comment (uilianries) [reply libraries/protobuf/serialization/main.py]: You are right :man_facepalming:

Comment (uilianries) [reply libraries/protobuf/serialization/main.py]: done

File: features/lockfiles/ci/build.py (hunk 6 above)
Comment (uilianries) [reply features/lockfiles/ci/build.py]: fixed by https://github.com/conan-io/examples/pull/29

File: features/emscripten/build.bat (hunk 1 above)
Comment (uilianries) [reply features/emscripten/build.bat]: done

**Comment by uilianries** [comment]
@jgsogo all green and updated!

**Comment by jgsogo** [comment]
Why [this last change](https://github.com/conan-io/examples/pull/25/commits/6c713d72e5dcb84426d445eac8a427c2a44d9c91)? It was already working in the previous changeset.

**Comment by jgsogo** [comment]
It looks like the environment variable for `CONAN_USER_HOME` is not working... 🤷‍♂

**Comment by uilianries** [comment]
:sob: 

checking it now

**Comment by jgsogo** [comment]
Let's merge it before it fails again 🙈 😆

**Comment by uilianries** [comment]
Success!

**Pull request merged**